engine/execution/ingestion/throttle.go

```diff
@@ -0,0 +1,241 @@
+// blocks until the execution has caught up
+const CatchUpThreshold = 500
+
+// BlockThrottle is a helper struct that helps throttle the unexecuted blocks to be sent
+// to the block queue for execution.
+// It is useful for case when execution is falling far behind the finalization, in which case
+// we want to throttle the blocks to be sent to the block queue for fetching data to execute
+// them. Without throttle, the block queue will be flooded with blocks, and the network
+// will be flooded with requests fetching collections, and the EN might quickly run out of memory.
+type BlockThrottle struct {
+	// config
+	threshold int // catch up threshold
+
+	// state
+	mu        sync.Mutex
+	executed  uint64
+	finalized uint64
+	inited    bool
+
+	// notifier
+	processables chan<- flow.Identifier
+
+	// dependencies
+	log     zerolog.Logger
+	state   protocol.State
+	headers storage.Headers
+}
+
+func NewBlockThrottle(
+	log zerolog.Logger,
+	state protocol.State,
+	execState state.ExecutionState,
+	headers storage.Headers,
+	catchupThreshold int,
+) (*BlockThrottle, error) {
+	finalizedHead, err := state.Final().Head()
+	if err != nil {
+		return nil, fmt.Errorf("could not get finalized head: %w", err)
+	}
+
+	finalized := finalizedHead.Height
+	// TODO: implement GetHighestFinalizedExecuted for execution state when storehouse
+	// is not used
+	executed := execState.GetHighestFinalizedExecuted()
+
+	if executed > finalized {
+		return nil, fmt.Errorf("executed finalized %v is greater than finalized %v", executed, finalized)
+	}
+
+	return &BlockThrottle{
+		threshold: catchupThreshold,
+		executed:  executed,
+		finalized: finalized,
+
+		log:     log.With().Str("component", "block_throttle").Logger(),
+		state:   state,
+		headers: headers,
+	}, nil
+}
+
+func (c *BlockThrottle) Init(processables chan<- flow.Identifier) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	c.log.Info().Msgf("initializing block throttle")
+	if c.inited {
+		return fmt.Errorf("throttle already inited")
+	}
+
+	c.inited = true
+	c.processables = processables
+
+	var unexecuted []flow.Identifier
+	var err error
+	if caughtUp(c.executed, c.finalized, c.threshold) {
+		unexecuted, err = findAllUnexecutedBlocks(c.state, c.headers, c.executed, c.finalized)
+		if err != nil {
+			return err
+		}
+		c.log.Info().Msgf("loaded %d unexecuted blocks", len(unexecuted))
+	} else {
+		unexecuted, err = findFinalized(c.state, c.headers, c.executed, c.executed+500)
+		if err != nil {
+			return err
+		}
+		c.log.Info().Msgf("loaded %d unexecuted finalized blocks", len(unexecuted))
+	}
+
+	for _, id := range unexecuted {
+		c.processables <- id
+	}
+
+	c.log.Info().Msgf("throttle initialized with %d unexecuted blocks", len(unexecuted))
+
+	return nil
+}
+
+func (c *BlockThrottle) OnBlockExecuted(_ flow.Identifier, executed uint64) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	if !c.inited {
+		return fmt.Errorf("throttle not inited")
```

Comment:
This is to make sure `c.processables` is not `nil`, which is supposed to be updated already by the `Init` function. If we continue, it will panic with nil error. 

But I will refactor to turn `inited` as a function and just check `if c.processables == nil`.